src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.ts

```diff
@@ -0,0 +1,193 @@
+            const participation = this.exercise.participations[0];
+            if (
+                participation.initializationState === InitializationState.INITIALIZED ||
+                participation.initializationState === InitializationState.FINISHED
+            ) {
+                return this.exercise.type === ExerciseType.MODELING ? ParticipationStatus.MODELING_EXERCISE : ParticipationStatus.TEXT_EXERCISE;
+            }
+        }
+
+        if (!this.hasParticipations(this.exercise)) {
+            return ParticipationStatus.UNINITIALIZED;
+        } else if (this.exercise.participations[0].initializationState === InitializationState.INITIALIZED) {
+            return ParticipationStatus.INITIALIZED;
+        }
+        return ParticipationStatus.INACTIVE;
+    }
+
+    hasParticipations(exercise: Exercise): boolean {
+        return exercise.participations && exercise.participations.length > 0;
+    }
+
+    hasResults(participation: Participation): boolean {
+        return participation.results && participation.results.length > 0;
+    }
+
+    isPracticeModeAvailable(): boolean {
+        const quizExercise = this.exercise as QuizExercise;
+        return quizExercise.isPlannedToStart && quizExercise.isOpenForPractice && moment(quizExercise.dueDate).isBefore(moment());
+    }
+
+    isOnlineEditorAllowed(): boolean {
+        return (this.exercise as ProgrammingExercise).allowOnlineEditor;
+    }
+
+    publishBuildPlanUrl(): boolean {
+        return (this.exercise as ProgrammingExercise).publishBuildPlanUrl;
+    }
+
+    onCopyFailure() {
+        console.log('copy fail!');
+    }
+
+    onCopySuccess() {
+        this.wasCopied = true;
+        setTimeout(() => {
+            this.wasCopied = false;
+        }, 3000);
+    }
+
+    startExercise() {
+        this.exercise.loading = true;
+
+        if (this.exercise.type === ExerciseType.QUIZ) {
+            // Start the quiz
+            return this.router.navigate(['/quiz', this.exercise.id]);
+        }
+
+        this.courseExerciseService
+            .startExercise(this.courseId, this.exercise.id)
+            .finally(() => (this.exercise.loading = false))
+            .subscribe(
+                participation => {
+                    if (participation) {
+                        this.exercise.participations = [participation];
+                        this.exercise.participationStatus = this.participationStatus();
+                    }
+                    if (this.exercise.type === ExerciseType.PROGRAMMING) {
+                        this.jhiAlertService.success('arTeMiSApp.exercise.personalRepository');
+                    }
+                },
+                error => {
+                    console.log('Error: ' + error);
+                    this.jhiAlertService.warning('arTeMiSApp.exercise.startError');
+                }
+            );
+    }
+
+    buildSourceTreeUrl(cloneUrl: string): string {
```

Comment:
We already have functionality for building the url and similar stuff like receiving the password somewhere.
Is there a way of reusing existing functionality?

Reply:
Maybe we can create a Service for this.

Reply:
the other usage will probably be removed at some point but a service is possible of course